### PR TITLE
[PyROOT] Move DeclareCppCallable to Numba.Declare

### PR DESCRIPTION
**Pull request by stwunsch**
* Rename DeclareCppCallable to Numba.Declare
* Remove generic backend
* Rename and adapt test accordingly
* Add support for RVecs in callables jitted with numba
* Add tutorial

~We desperately need numba on the PR build nodes!~ Taken care of!

### CODE REVIEW

**Comment by stwunsch** [comment]
The PR depends on #5285 to implement the `Numba` namespace lazily.

**Comment by stwunsch** [comment]
@phsft-bot build

**Comment by stwunsch** [comment]
@phsft-bot build

**Comment by stwunsch** [comment]
@phsft-bot build

**Comment by stwunsch** [comment]
@phsft-bot build

**Comment by stwunsch** [comment]
@phsft-bot build

**Comment by stwunsch** [comment]
@phsft-bot build

**Comment by eguiraud** [comment]
Hi, I'll add a review within today

File: bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+        # Create Python wrapper with C++ friendly signature
+
+        # Define signature
+        pywrapper_signature = [
+                'ptr_{0}, size_{0}'.format(i) if 'RVec' in t else 'x_{}'.format(i) \
+                        for i, t in enumerate(input_types)]
+        if 'RVec' in return_type:
+            # If we return an RVec, we return via pointer the pointer of the allocated data,
+            # the size in elements. In addition, we provide the size of the datatype in bytes.
+            pywrapper_signature += ['ptrptr_r, ptrsize_r', 'dtypesize_r']
```
Comment (eguiraud) [inline bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py:112]: Could the logic be simplified a bit by removing the `dtypesize_r` argument from the signature and just use the value we know it has (because numba types are always strictly "sized")?

File: bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+                'memcpy(x_r.data(), ptr, size * dtype_size);',
+                'free(ptr);',
+                # If we return a RVec<bool>, we rely here on the automatic conversion of RVec<char> to RVec<bool>
+                'return x_r;'])
+        else:
+            return_op = 'return funcptr({});'.format(', '.join(func_args))
+
+        # Build wrapper code
+        cppwrappercode = """\
+#pragma cling optimize(3)
```
Comment (eguiraud) [inline bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py:248]: this does not do anything right?

File: bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+                VECBOOL_CONVERSION='\n    '.join(vecbool_conversion),
+                RETURN_OP=return_op)
+
+        # Jit wrapper C++ code
+        err = gbl_namespace.gInterpreter.Declare(cppwrappercode)
+        if not err:
+            raise Exception('Failed to jit C++ wrapper code with cling:\n{}'.format(cppwrappercode))
+        func.__cpp_wrapper__ = cppwrappercode
+
+        return func
```
Comment (eguiraud) [inline bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py:276]: should we return the numba-jitted function instead? Is it less surprising if `@Numba.Declare` returns a numba-jitted function or the function itself?

File: bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+
+        # Define return operation
+        if 'RVec' in return_type:
+            innert = get_inner_type(return_type)
+            if innert == 'bool': innert = 'char'
+            return_op = '\n    '.join([
+                '// Because an RVec cannot take the ownership of external data, we have to copy the returned array',
+                'long size; // Size of the returned array',
+                '{}* ptr; // Pointer to the data of the returned array'.format(innert),
+                'int dtype_size = sizeof({}); // Size of array datatype in bytes'.format(innert),
```
Comment (eguiraud) [inline bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py:236]: nitpicking, should be `std::size_t` instead of `int`

File: bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+            glob['dtype_r'] = get_numba_type(get_inner_type(return_type))
+
+        exec(pywrappercode, glob, locals())
+        if not 'pywrapper' in locals():
+            raise Exception('Failed to create Python wrapper function:\n{}'.format(pywrappercode))
+
+        # Jit the Python wrapper code
+        c_return_type, c_input_types = get_c_signature(input_types, return_type)
+        try:
+            nbcfunc = nb.cfunc(c_return_type(*c_input_types), nopython=True, inline='always')(locals()['pywrapper'])
```
Comment (eguiraud) [inline bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py:172]: i don't think it actually hurts, but we don't really need the `inline='always'` here

File: bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py (hunk 4 above)
Comment (stwunsch) [reply bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py:236]: Correct :) Other questions: Is `long` enough for the size or should we go do `unsigned long`? Or even `unsigned long long`? Or just a `size_t`? :D

File: bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py (hunk 5 above)
Comment (stwunsch) [reply bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py:172]: True, shouldn't change anything. I left it there for completeness.

File: bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py (hunk 3 above)
Comment (stwunsch) [reply bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py:276]: Hmm, ... I don't know!

File: bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py (hunk 2 above)
Comment (stwunsch) [reply bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py:248]: Yes, right. As we learnt in the PPP, the code generation does actually not happen.

Comment (stwunsch) [reply bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py:248]: But it does not hurt :D

File: bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py (hunk 1 above)
Comment (stwunsch) [reply bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py:113]: I get the type via a `sizeof` call from C++, but you are also right, we could go through numba and get the size of the type from them. I check whether this is easily possible because I would like to prevent another map!

File: bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+    except:
+        raise Exception('Failed to import numba')
+    try:
+        import cffi
+    except:
+        raise Exception('Failed to import cffi')
+    try:
+        import re
+    except:
+        raise Exception('Failed to re')
```
[the remark this answers is not in the document]
Comment (stwunsch) [reply bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py:17]: No, we should remove it.

File: bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py (hunk 4 above)
Comment (eguiraud) [reply bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py:237]: we should go with the type of `sizeof`, `std:size_t`

File: bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py (hunk 5 above)
Comment (eguiraud) [reply bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py:173]: it's a bit of extra mental load for the reader because now you are thinking you did not understand the logic because you don't expect this function to be inlined in another numba function

File: bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py (hunk 2 above)
Comment (eguiraud) [reply bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py:249]: it's misleading, can you remove it please?

File: bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py (hunk 1 above)
Comment (eguiraud) [reply bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py:113]: i'd say even another map would be a simplification w.r.t. generating code that conditionally has the extra argument and then generating code that conditionally passes the argument after calling sizeof(...) from jitted C++ :sweat_smile:

Comment (stwunsch) [reply bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py:113]: but you have to maintain the map, not the logic :) though, i understand your point

Comment (eguiraud) [reply bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py:113]: true, up to you

**Comment by eguiraud** [comment]
Sorry for the multiple repeated comments, github was glitching hard yesterday, i have deleted them now

File: bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+        c_input_types = []
+        for t in input_types:
+            if 'RVec' in t:
+                c_input_types += [nb.types.CPointer(get_numba_type(get_inner_type(t))), nb.int32]
+            else:
+                c_input_types.append(get_numba_type(t))
+        if 'RVec' in return_type:
+            # See the Python wrapper for the reason using these types
+            c_return_type = nb.void
+            c_input_types += [nb.types.CPointer(nb.int64), nb.types.CPointer(nb.int64), nb.int32]
```
[the remark this answers is not in the document]
Comment (stwunsch) [reply bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py:72]: We cannot cast the pointer in the python impl and it's always a int64 there, so we return it here and cast it in c++.

File: bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+                }
+        if t in typemap:
+            return typemap[t]
+        raise Exception(
+                'Type {} is not supported for jitting with numba. Valid fundamental types and RVecs thereof are {}'.format(
+                    t, list(typemap.keys())))
+
+    def get_c_signature(input_types, return_type):
+        '''
+        Get C signature as numba type objects from C++ typenames
```
[the remark this answers is not in the document]
Comment (stwunsch) [reply bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py:60]: Good point, also extended the signature of the `get_numba_types`.

File: bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py (hunk 1 above)
Comment (stwunsch) [reply bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py:113]: I tried to get in a `numba.float32.bitwidth` or similar. Unfortuantley numba doesn't recognize the type / the member function in the jitting with `nopython=True`. So i'll stick to this solution :/

File: bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py (hunk 2 above)
Comment (stwunsch) [reply bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py:249]: Done

File: bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py (hunk 4 above)
Comment (stwunsch) [reply bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py:237]: we cast it anyway to an `int` when we call the function. and which type should be larger than an `int`?

File: bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py (hunk 5 above)
Comment (stwunsch) [reply bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py:173]: removed

File: bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py (hunk 1 above)
Comment (eguiraud) [reply bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py:113]: aaah `bitwidth`, well done finding that one, I couldn't when I looked for it! You should not call `numba.float32.bitwidth` from the jitted function -- you can just inject the actual value in the generated code. that simplifies the logic quite a bit

File: bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py (hunk 3 above)
Comment (eguiraud) [reply bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py:277]: i think either solution is fine as long as the documentation of `ROOT.Numba.Declare` says what the decorator does to the function (either nothing, or it transforms it into a numba function).

Which makes me think: are there any user docs for this feature?

File: bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py (hunk 4 above)
Comment (eguiraud) [reply bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py:237]: alright!

File: bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (eguiraud) [reply bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py:72]: alright, let me check that I understand the type punning that's going on:

* C++ has a `T* ptr`
* it passes it to a numba-jitted cfunc as `&ptr`
* the numba cfunc converts `T**` to `long int*`
* the numba cfunc makes a memcpy from the `T*` returned by the numba-jitted user function to a malloc'd `void *`
* the numba cfunc assigns `*longintptr = mallocd_voidptr`

Is that correct? If yes I'd like to ask on PPP whether anyone thinks there is anything illegal/potentially broken there

File: bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py (hunk 3 above)
Comment (stwunsch) [reply bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py:277]: There's a tutorial :D And I'll add the numba jitted function as an attribute!

File: bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py (hunk 1 above)
Comment (stwunsch) [reply bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py:113]: ah i'm stupid :see_no_evil:

File: bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py (hunk 3 above)
Comment (stwunsch) [reply bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py:277]: Now you can get from the callable `f` the numba jitted func as `f.jitted`. Any better suggestions for the wording?

Comment (stwunsch) [reply bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py:277]: Probably, `f.jitted_func`?

File: bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py (hunk 1 above)
Comment (stwunsch) [reply bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py:113]: Thanks, works :dancing_men:

File: bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py (hunk 3 above)
Comment (eguiraud) [reply bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py:277]: > There's a tutorial

There is a discoverability issue (users don't go look for a tutorial for a feature they don't know), but it's not for this PR to solve it. Maybe we should have a "Using RDF from python" section in the RDF user guide that points to things like `Numba.Declare`.

`f.jitted[_func]` could refer to either numba's jitted function or cling's. I'd suggest  something like `f.numba_func`

File: bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py (hunk 1 above)
Comment (eguiraud) [reply bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py:113]: wo-hoo! :smile:

File: bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+
+    # Define inner decorator without arguments
+    def inner(func, input_types=input_types, return_type=return_type, name=name):
+
+        # Jit the given Python callable with numba
+        nb_return_type, nb_input_types = get_numba_signature(input_types, return_type)
+        try:
+            nbjit = nb.jit(nb_return_type(*nb_input_types), nopython=True, inline='always')(func)
+        except:
+            raise Exception('Failed to jit Python callable {} with numba.jit'.format(func))
```
[the remark this answers is not in the document]
Comment (stwunsch) [reply bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py:127]: It's good to see the pointer so you can debug whether you actually see the function you expect. Since you can replace in Python an object with a given name at any time, a pointer is the only id you have.

File: bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+
+    input_types = [normalize_typename(t) for t in input_types]
+    return_type = normalize_typename(return_type)
+
+    # Helper functions to determine types
+    def get_inner_type(t):
+        '''
+        Get inner typename of a templated C++ typename
+        '''
+        g = re.match('(.*)<(.*)>', t).groups(0)
```
[the remark this answers is not in the document]
Comment (stwunsch) [reply bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py:51]: Good idea, added try/catch guards.

File: bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (eguiraud) [reply bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py:127]: :+1:

File: bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py (hunk 3 above)
Comment (stwunsch) [reply bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py:303]: Taken!

File: bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (stwunsch) [reply bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py]: Let me rephrase and we'll have a look whether we agree that we see the same thing. I'm going from the inner most layer to the outer most:

1. In the jitted Python callable (not the c-friendly wrapper!) we just return the numpy/numba array.
2. Then we get the return value in the Python (c-friendly) wrapper. There we get the pointer to the data via `array.ctypes.data` as an `int64`. That value is returned to C++ via a `int64*`.
3. The `int64*` is implicitely converted by the function call to a `T**`, so by setting `ptr[0] = int64 value` in Python we get an implicit conversion of `int64` to `T*`.
4. Finally, we use the pointer in `malloc`, which takes a `void*` as input.

[the remark this answers is not in the document]
Comment (stwunsch) [reply bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py]: yep, i think we understand it the same way.

File: bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py (hunk 3 above)
Comment (stwunsch) [reply bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py:307]: Taken the `f.numba_func` and also added a rather massiv docstring to the outer decorator.

File: bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py [hunk 11]
```diff
@@ -0,0 +1,309 @@
...
+def _NumbaDeclareDecorator(input_types, return_type, name=None):
+    '''
+    Decorator for making Python callables accessible in C++ by just-in-time compilation
+    with numba and cling
+
+    The decorator takes the given Python callable and just-in-time compiles (jits)
+    wrapper functions with the given C++ types for input and return types. Eventually,
+    the Python callable is accessible in the Numba namespace in C++.
+
+    The implementation first jits with numba the Python callable. We support fundamental types
```
Comment (eguiraud) [inline bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py:13]: typo: fundamental types _and_

File: bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py (hunk 3 above)
Comment (eguiraud) [reply bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py:307]: great!

**Comment by stwunsch** [comment]
@etejedor Could you review the Python integration of the feature? You don't have to review all the nitty gritty details of the impl but the infrastructure.

File: bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py [hunk 12]
```diff
@@ -0,0 +1,309 @@
+from cppyy import gbl as gbl_namespace
```
Comment (etejedor) [inline bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py:1]: We can leave this file here, for now. But perhaps at some point we will need to reorganize. Everything that is not a pythonization but belongs more to the interface of the ROOT module, could go in an e.g. `interface` submodule (containing now `_facade.py`, `_numbadeclare.py`, `_application.py`).
